### ISSUE TITLE
RCP Simulation

### ISSUE THREAD

**Issue by ChengjunJia** [kaist-ina/ns2-xpass#8]
Hi, Mr. InhoCho. I am trying to use RCP and DCTCP in Fat-Tree topology with WebServer traffic load.
For DCTCP, I use the open-source from Alibaba's HPCC project (based on ns-3), it works fine.
But for RCP, I found that retransmission happened even when I set queueSize 1000,000 packets. Do you meet the problem when you used RCP?
BTW, I do not find how you use the RCP algorithm in this project? I will appreciate it if you share your RCP script.

Thanks a lot
Chengjun

**Comment by inhocho89** [comment]
Hi Chengjun,

I think one possibility causing retransmission is that the minimum rate of
RCP or granularity of the rate adjustment is too high (i.e., you are using
an integer type variable for rate, etc.).
For now, I cannot tell more detail as I don't know your experiment setup
(which codebase you are using, link speed, etc.)
For ExpressPass paper, we used an internal implementation of RCP which is
not open to the public, and unfortunately, we don't have any plan to make
it public any time soon.

I hope this information helps.

Best,
Inho

**Comment by ChengjunJia** [comment]
Thanks for your suggestion. I am setting all links as 100Gbps/1us, based on Professor Nick Mckeown's [CS224 class](https://reproducingnetworkresearch.wordpress.com/page/2/). The codes are running in one ns2 Docker Image.

I just doubted whether the TailDrop was due to the RCP design: based on rate control instead of congestion window. Now that in your experiment, you found that there was no packet TailDrop, I would try to debug it.

Thanks again.

**Comment by ChengjunJia** [comment]
I figure out that RCP would do retransmission if there is no response in 2RTT by default: `rto_timer_.resched(2*rtt_);`
So retransmission happens but there is no TailDrop if I set the queue much large (e.g. 10MB).

I found that if the queue was 1MB, it would encounter several tail drops but not many. I think it is reasonable for RCP with rate control would easily meet a high queue before it informs the sender to reduce the rate successfully.

Thanks a lot. I will close the issue.
